docs/zephyr: Add documentation for the Zephyr port

Julia-Hathaway opened the pull request:
Includes a draft quickref, tutorial, and port specific classes. 

Currently a work in progress, looking for some direction and initial review.

Julia-Hathaway commented:
@dpgeorge any feedback so far? Still a work in progress, but wanted to check in for some direction.

dpgeorge commented on docs/library/zephyr.DiskAccess.rst:
```diff
@@ -0,0 +1,29 @@
+.. _zephyr.DiskAccess:
+
+class DiskAccess -- access to disk storage
+===============================================
```
please make the underline the same length as the title

dpgeorge commented on docs/library/zephyr.DiskAccess.rst:
```diff
@@ -0,0 +1,29 @@
+class DiskAccess -- access to disk storage
+===============================================
+
+Uses zephyr disk access API.
```
would be good to provide a little bit of info about how disk access differs to flash map (eg what happens at the lower level, when to use one or the other)

dpgeorge commented on docs/library/zephyr.DiskAccess.rst:
```diff
@@ -0,0 +1,29 @@
+
+.. class:: DiskAccess(disk_name)
+
+   Gets an object for accessing disk memory of the specific disk.
```
can you provide some example names for `disk_name`?

dpgeorge commented on docs/library/zephyr.FlashArea.rst:
```diff
@@ -0,0 +1,29 @@
+class FlashArea -- access to built-in flash storage
+===================================================
+
+Uses zephyr flash map API.
```
as above, would be good to add some info about how flash map differs from disk access

dpgeorge commented on docs/library/zephyr.FlashArea.rst:
```diff
@@ -0,0 +1,29 @@
+
+.. class:: FlashArea(id)
+
+   Gets an object for accessing flash memory at partition specified by *id*.
```
as above, what are some example *id* values?

dpgeorge commented on docs/library/zephyr.rst:
```diff
@@ -0,0 +1,18 @@
+
+The ``zephyr`` module contains functions and classes specific to the Zephyr port.
+
+Classes
```
above this subsection you can add a subsection for the functions in `zephyr`: `is_preempt_thread`, `current_tid`, etc

dpgeorge commented on docs/zephyr/general.rst:
```diff
@@ -0,0 +1,22 @@
+
+The Zephyr Project is a Linux Foundation hosted Collaboration Project. It’s an open 
+source collaborative effort uniting developers and users in building a best-in-class 
+small, scalable, real-time operating system (RTOS) optimized for resource-constrained 
```
try not to use advertising wording (like "best in class"), it should be informative only

(yes it's a little subjective what's informative or not!)

dpgeorge commented on docs/zephyr/quickref.rst:
```diff
@@ -0,0 +1,151 @@
+   general.rst
+   tutorial/index.rst
+
+Running Micropython
```
please use MicroPython (capital P)

dpgeorge commented on docs/zephyr/quickref.rst:
```diff
@@ -0,0 +1,151 @@
+    switch.irq(lambda t: print("SW2 changed"))
+    SW3.irq(lambda t: print("SW3 changed"))
+
+    ...
```
no need for these dots

dpgeorge commented on docs/zephyr/quickref.rst:
```diff
@@ -0,0 +1,151 @@
+    disk = DiskAccess(*disk_name*)      # create disk object
+    print(disk)                         # print disk name
+
+    disk.readblocks(*block num*, *buffer*, *offset*)
```
would be better to give examples of how to format and mount a filesystem, using `uos.VfsLfs2.mkfs` etc

dpgeorge commented on docs/zephyr/tutorial/pins.rst:
```diff
@@ -0,0 +1,33 @@
+
+        LED = Pin(("GPIO_1", 22), Pin.OUT)
+        
+Reference your board's datasheet or Zephyr documentation for pin numbers.
```
The pin labelling is hard to get right.  I think it'd be good to provide more info here, eg for a selection of boards what the format/style is for the pin names/numbers.

dpgeorge commented:
Thanks for this, it's a really good start and I think a good addition now that the zephyr port has more functionality.  I've left some more specific comments above.

Julia-Hathaway committed:
docs/zephyr: Add quick reference for the Zephyr port.

Includes an introduction to using the Zephyr port on Micropython.
The quickref details examples of how to use each module the port
currently supports.
The tutorial provides additional details for Zephyr specific modules.

Signed-off-by: Julia Hathaway <julia.hathaway@nxp.com>

Julia-Hathaway committed:
docs/library/zephyr: Add libraries specific to the Zephyr port.

Includes documentation for Zephyr specific modules (zephyr and zsensor),
classes (DiskAccess and FlashArea), and functions.

Signed-off-by: Julia Hathaway <julia.hathaway@nxp.com>

MaureenHelm reviewed (approved):
This is a great start! There are a few areas that could be expanded, in particular could use some discussion about how Kconfig settings affect which MicroPython features are available, but that can be done in a future PR.

Julia-Hathaway commented:
@dpgeorge any feedback on these additions?

peterhinch commented:
Apologies for the OT message but a [forum user](https://forum.micropython.org/viewtopic.php?f=15&t=10945&hilit=zephyr) has been struggling to set up a UART on Zephyr. It seems as if the **.init** method is missing. Perhaps one of you Zephyr gurus might like to help.

MaureenHelm commented:
> @dpgeorge any feedback on these additions?

ping

dpgeorge commented on docs/library/zephyr.zsensor.rst:
```diff
@@ -0,0 +1,123 @@
+
+    Obtains a measurement sample from the sensor device using Zephyr sensor_sample_fetch and
+    stores it in an internal driver buffer as a useful value, a tuple of {integer part of value,
+    fractional part of value in 1-millionths}.
```
It might be confusing using braces `{..}` for a "tuple" in this context (because in Python `{}` is used for dict/set, and tuple means `(a, b, ...)`).  Maybe call it "pair" here and use normal parenthesis `()`?

dpgeorge commented on docs/library/zephyr.zsensor.rst:
```diff
@@ -0,0 +1,123 @@
+
+.. data:: PROX
+
+    Proximity. Adimensional. A value of 1 indicates that an object is close.
```
Adimensional -> dimensionless

dpgeorge commented:
> any feedback on these additions?

Looks great!

Please address the few outstanding comments from @MaureenHelm  and myself.

Also, it might be worth mentioning somewhere in the tutorial (eg in the storage section) that if you have a filesystem and create a `main.py` file in the root of it, then that will be executed automatically on start up.

dpgeorge commented:
Thanks for resolving the review comments.  Rebased and merged in 333e16521bc8af8a55f362c1bad2f21c4704d20b and 6ed69906257ee59d6533dc06ddfa02db5f4fbce3, with follow-up in 671f012306e0942cbe59fcd43e25e67c289da351
